tokio-util/src/codec/mod.rs

```diff
@@ -1,18 +1,255 @@
 //! Adaptors from AsyncRead/AsyncWrite to Stream/Sink
 //!
-//! Raw I/O objects work with byte sequences, but higher-level code
-//! usually wants to batch these into meaningful chunks, called
-//! "frames".
+//! Raw I/O objects work with byte sequences, but higher-level code usually
+//! wants to batch these into meaningful chunks, called "frames".
 //!
-//! This module contains adapters to go from streams of bytes,
-//! [`AsyncRead`] and [`AsyncWrite`], to framed streams implementing
-//! [`Sink`] and [`Stream`].  Framed streams are also known as
-//! transports.
+//! This module contains adapters to go from streams of bytes, [`AsyncRead`] and
+//! [`AsyncWrite`], to framed streams implementing [`Sink`] and [`Stream`].
+//! Framed streams are also known as transports.
+//!
+//! # The Decoder trait
+//!
+//! A [`Decoder`] is used together with [`FramedRead`] or [`Framed`] to turn an
+//! [`AsyncRead`] into a [`Stream`]. The job of the decoder trait is to specify
+//! how sequences of bytes are turned into a sequence of frames, and to
+//! determine where the boundaries between frames are.  The job of the
+//! `FramedRead` is to repeatedly switch between reading more data from the IO
+//! resource, and asking the decoder whether we have received enough data to
+//! decode another frame of data.
+//!
+//! The main method on the `Decoder` trait is the [`decode`] method. This method
+//! takes as argument the data that has been read so far, and when it is called,
+//! it will be in one of the following situations:
+//!
+//!  1. The buffer contains less than a full frame.
+//!  2. The buffer contains exactly a full frame.
+//!  3. The buffer contains more than a full frame.
+//!
+//! In the first situation, the decoder should return `Ok(None)`.
+//!
+//! In the second situation, the decoder should clear the provided buffer and
+//! return `Ok(Some(the_decoded_frame))`.
+//!
+//! In the third situation, the decoder should use a method such as [`split_to`]
+//! or [`advance`] to modify the buffer such that the frame is removed from the
+//! buffer, but any data in the buffer after that frame should still remain in
+//! the buffer. The decoder should also return `Ok(Some(the_decoded_frame))` in
+//! this case.
+//!
+//! Finally the decoder may return an error if the data is invalid in some way.
+//! The decoder should _not_ return an error just because it has yet to receive
+//! a full frame.
+//!
+//! It is guaranteed that, from one call to `decode` to another, the provided
+//! buffer will contain the exact same data as before, except that if more data
+//! has arrived through the IO resource, that data will have been appended to
+//! the buffer.  This means that reading frames from a `FramedRead` is
+//! essentially equivalent to the following loop:
```

Comment:
```suggestion
//! essentially equivalent to the following loop:
//!
```

Newline makes this easier to read imo